terminal.go

```diff
@@ -449,9 +456,21 @@ func visualLength(runes []rune) int {
 	return length
 }
 
+// handleOnKey checks if the given key has a OnKey function assigned to it
+// which, if existent, gets called. Returns false if the corresponding OnKey
+// function returns also false, on all other cases true
+func (t *Terminal) handleOnKey(key rune) bool {
+	f, ok := t.OnKey[key]
+	return !ok || ok && f(key)
+}
+
 // handleKey processes the given key and, optionally, returns a line of text
 // that the user has entered.
 func (t *Terminal) handleKey(key rune) (line string, ok bool) {
+	if !t.handleOnKey(key) {
+		return
+	}
+
```

Comment:
Instead of inserting the check here as well as sprinkled judiciously throughout `readLine` (a little hard to follow), consider inserting it in one common location. I haven't verified, but immediately after checking for `utf8.RuneError`, around line 759, seems like it would catch all these same cases.